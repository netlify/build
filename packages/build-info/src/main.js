import { getContext } from './context.js'
import { buildInfo } from './core.js'

export const getBuildInfo = async function (opts) {
  const context = await getContext(opts)
  return await buildInfo(context)
}
