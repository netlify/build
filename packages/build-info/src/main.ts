import { ContextOptions, getContext } from './context.js'
import { buildInfo } from './core.js'

export const getBuildInfo = async (opts: ContextOptions) => {
  const context = await getContext(opts)
  return await buildInfo(context)
}
