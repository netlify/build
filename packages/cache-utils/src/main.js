'use strict'

require('./utils/polyfills')

const del = require('del')

const { getCacheDir } = require('./dir')
const { moveCacheFile, hasFiles } = require('./fs')
const { list } = require('./list')
const { getManifestInfo, writeManifest, removeManifest, isExpired } = require('./manifest')
const { parsePath } = require('./path')

// Cache a file
const saveOne = async function (
  path,
  { move = DEFAULT_MOVE, ttl = DEFAULT_TTL, digests = [], cacheDir, cwd: cwdOpt, mode } = {},
) {
  const { srcPath, cachePath } = await parsePath({ path, cacheDir, cwdOpt, mode })

  if (!(await hasFiles(srcPath))) {
    return false
  }

  const { manifestInfo, identical } = await getManifestInfo({ cachePath, move, ttl, digests })
  if (identical) {
    return true
  }

  await del(cachePath, { force: true })
  await moveCacheFile(srcPath, cachePath, move)
  await writeManifest(manifestInfo)

  return true
}

// Restore a cached file
const restoreOne = async function (path, { move = DEFAULT_MOVE, cacheDir, cwd: cwdOpt, mode } = {}) {
  const { srcPath, cachePath } = await parsePath({ path, cacheDir, cwdOpt, mode })

  if (!(await hasFiles(cachePath))) {
    return false
  }

  if (await isExpired(cachePath)) {
    return false
  }

  await del(srcPath, { force: true })
  await moveCacheFile(cachePath, srcPath, move)

  return true
}

// Remove the cache of a file
const removeOne = async function (path, { cacheDir, cwd: cwdOpt, mode } = {}) {
  const { cachePath } = await parsePath({ path, cacheDir, cwdOpt, mode })

  if (!(await hasFiles(cachePath))) {
    return false
  }

  await del(cachePath, { force: true })
  await removeManifest(cachePath)

  return true
}

// Check if a file is cached
const hasOne = async function (path, { cacheDir, cwd: cwdOpt, mode } = {}) {
  const { cachePath } = await parsePath({ path, cacheDir, cwdOpt, mode })

  return (await hasFiles(cachePath)) && !(await isExpired(cachePath))
}

const DEFAULT_MOVE = false
const DEFAULT_TTL = undefined

// Allow each of the main functions to take either a single path or an array of
// paths as arguments
const allowMany = async function (func, paths, ...args) {
  if (!Array.isArray(paths)) {
    return func(paths, ...args)
  }

  const results = await Promise.all(paths.map((path) => func(path, ...args)))
  return results.some(Boolean)
}

const save = allowMany.bind(null, saveOne)
const restore = allowMany.bind(null, restoreOne)
const remove = allowMany.bind(null, removeOne)
const has = allowMany.bind(null, hasOne)

// Change `opts` default values
const bindOpts = function (opts) {
  return {
    save: (paths, optsA) => save(paths, { ...opts, ...optsA }),
    restore: (paths, optsA) => restore(paths, { ...opts, ...optsA }),
    remove: (paths, optsA) => remove(paths, { ...opts, ...optsA }),
    has: (paths, optsA) => has(paths, { ...opts, ...optsA }),
    list: (optsA) => list({ ...opts, ...optsA }),
    getCacheDir: (optsA) => getCacheDir({ ...opts, ...optsA }),
  }
}

module.exports = { save, restore, remove, has, list, getCacheDir, bindOpts }
