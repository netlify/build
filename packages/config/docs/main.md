# Netlify configuration

## Purpose

The goal of `@netlify/config` is to abstract the logic that loads, normalizes and validates the Netlify configuration
(`netlify.toml`, UI build settings, `_headers`, `_redirects`, etc.). It returns a single normalized object to consumers.

Note: while `@netlify/build` share the same monorepo, it has a very distinct purpose.

## Consumers

At the moment, `@netlify/config` is used by four parts of Netlify's architecture. It might also be used by additional
consumers in the future, including outside of Netlify.

It is possible to distinguish between those different consumers using the [`mode` flag](../README.md#mode) which is set
to `"buildbot"`, `"cli"` or `"require"` by each of them. However, we try to limit the usage of that flag as much as
possible:

- Instead, we want those different modes to behave as similarly to each other as possible, especially Netlify CLI and
  buildbot.
- There are some cases where those different environments do require different logic. When this happens, we try to add
  new flags to toggle each specific behavior, as opposed to rely on the `mode` flag which would be less composable and
  re-usable.

### Programmatic vs CLI

There is both
[a CLI](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/bin/main.js#L16)
and
[programmatic entry point](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/main.js#L30).

To ensure both entry points re-use as much logic as possible, we try to minimize the amount of CLI-specific logic. At
the moment, the CLI code only includes some logic that makes sense exclusively within that context: exit code, CLI flags
parsing, terminal output.

By default, the CLI entry point prints to `stdout`. However,
[an `--output` flag](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/bin/main.js#L38)
is used
[by the buildbot](https://github.com/netlify/buildbot/blob/b6c6fc159f077f5dbda69ceb17363960232243fe/bot/configuration.go#L129)
to print to a file instead, which avoids crashing the process when the configuration object is very big.

### Buildbot

The buildbot logic uses multiple configuration properties: build directory, Functions directory, etc.Therefore,
[`@netlify/config` is called via the CLI](https://github.com/netlify/buildbot/blob/b6c6fc159f077f5dbda69ceb17363960232243fe/bot/configuration.go#L120)
(since the buildbot is written in a different language) at the beginning of each build.

`@netlify/config` is installed
[at Docker build time](https://github.com/netlify/buildbot/blob/b6c6fc159f077f5dbda69ceb17363960232243fe/package.json#L12)
by the buildbot. It is run with
[a hardcoded Node.js version](https://github.com/netlify/buildbot/blob/b6c6fc159f077f5dbda69ceb17363960232243fe/script/install-nvm.sh#L16).

### Netlify Build

`@netlify/build` also
[loads the configuration object](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/build/src/core/config.js#L107)
with `@netlify/config`, but programmatically.

Since the buildbot already loads that same object, we cache it
[as a `cachedConfig` object](https://github.com/netlify/buildbot/blob/b6c6fc159f077f5dbda69ceb17363960232243fe/script/run-build.sh#L75)
so `@netlify/build` does not need to re-compute it. This also ensures `@netlify/build` and the buildbot use the exact
same configuration properties.

#### Build plugins

Build plugins can both access and modify the Netlify configuration normalized object returned by `@netlify/config` using
[a `netlifyConfig` property passed to them](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/build/src/plugins/child/run.js#L16).

[When modified](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/build/src/plugins/child/diff.js#L21),
`@netlify/config`
[is called again](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/build/src/core/config.js#L126).
It is also called again when a build command or plugin
[adds a `_headers` or `_redirects` file](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/build/src/steps/update_config.js#L59).

The new configuration properties are printed, with some additional information when in debug mode.

The configuration mutations
[are stored in an `inlineConfig` object](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/mutations/apply.js#L14),
which is
[passed as a flag to `@netlify/config`](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/build/src/steps/update_config.js#L13)
and has the highest merging priority.

The logic in `@netlify/build` takes into account that the Netlify configuration can be mutated at any point. While most
configuration properties can be mutated,
[some are still read-only](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/mutations/apply.js#L52)
due to that additional complexity.

In production, any configuration mutation
[is temporarily persisted to `netlify.toml`](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/mutations/update.js#L17)
before deploying the site, and restored afterwards.

### Netlify CLI

Most Netlify CLI commands rely on configuration properties. The base command loads it
[using `@netlify/config` programmatically](https://github.com/netlify/cli/blob/f87d9e6e1749bedaabaace3dae98ca1ed6d84fd6/src/commands/base-command.js#L474).

An `inlineConfig` flag was added to `@netlify/config` so Netlify CLI can override specific configuration properties
based on its own CLI flags, but this is not used yet (see https://github.com/netlify/cli/issues/1265,
https://github.com/netlify/cli/issues/960, https://github.com/netlify/cli/issues/961).

We try to
[emulate the same environment variables](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/env/main.js#L13)
as on the buildbot, as much as possible.

### Automated tests

A few flags are available for programmatic users. Those are mostly used by our automated tests:

- [The `testOpts` flag](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/bin/flags.js#L153)
- [The `buffer` flag](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/bin/flags.js#L166)
  which allows returning the output programmatically instead of printing it on the terminal

## File paths

The Netlify configuration relies on many file paths. Those are documented [here](file_paths.md).

## UI build settings

Users can specify the configuration either in files (`netlify.toml`, `_headers`, `_redirects`) or in the Netlify UI
("build settings"). The latter has lower merging priority.

UI build settings are passed to `@netlify/config` using
[the `defaultConfig` flag](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/default.js#L6).

In the buildbot, this is populated
[with some values](https://github.com/netlify/buildbot/blob/7383bb8971c718de9618a700ec5049cb71af7102/bot/buildbot.go#L135-L144)
coming from the backend API.

When called from the Netlify CLI, API calls to the backend
[are being made](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/api/site_info.js#L24)
to retrieve
[those values](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/api/build_settings.js#L7).
This relies on multiple flags:

- The
  [`token`](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/api/client.js#L7)
  and
  [`siteId`](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/api/site_info.js#L30)
  flags require the site to have been linked locally and the user to be logged in.
- [The `host`, `scheme` and `pathPrefix` flags](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/api/client.js#L12)
  allow overriding the API endpoint URL.
- [The `offline` flag](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/api/client.js#L7)
  can disable this.

## Headers and redirects

Redirects configuration is more complicated:

- It has multiple properties
- It can be specified either in `netlify.toml` or in a separate `_redirects` file located in the publish directory and
  with its own Netlify-specific file format

Therefore, that logic is abstracted to
[its own module `netlify-redirect-parser`](https://github.com/netlify/netlify-redirect-parser).

Headers are handled very similarly to redirects, for consistency. They use a separate `_headers` file
[and `netlify-headers-parser` module](https://github.com/netlify/netlify-headers-parser).

## Context-specific configuration

Users can make some configuration specific to a given context or branch. Those are specified by embedding a
configuration object inside the main one, but namespaced with `context.{context|branch}.*`. `@netlify/config`
[merges those with higher priority](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/context.js#L28),
but only if the current context or branch matches.

Those are specified with:

- The `context` flag, which
  [defaults to `"production"`](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/options/main.js#L42)
- The `branch` flag,
  [which defaults](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/options/branch.js#L9)
  to the `BRANCH` environment variable, `HEAD` branch, `main` branch or `"master"` (this name was kept for backward
  compatibility).

## Normalization and validation

As described above, the final configuration object is a product of multiple merges:

- The configuration object loaded from:
  - Files (`config`)
  - The UI build settings (`defaultConfig`)
  - Build plugins modifications (`inlineConfig`)
- Context/branch-specific properties

As a result, normalization and validation happens
[in multiple stages](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/main.js#L259).
For example, we need to validate that `context` is a plain object before merging context properties. However, since the
build command might be specified either in `context` properties or not, we can only validate it after those have been
merged.

Additionally, we keep track of where the property was originally set
[with sibling properties named `origin`](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/origin.js#L11).
At the moment, this is mostly used for logging and debugging purposes. For example, the build logs show whether the
build command was set in `netlify.toml` or in the UI.

Almost all `@netlify/config` errors are user configuration errors. To help both users debug those problems and support
focus on actual system bugs, we enforce some strict configuration
[validation](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/validate/validations.js#L40).
Error messages are printed in the build logs with a nice format and examples on how to fix them. This logic relies on
declaring each configuration property's shape and example values
[in the following file](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/validate/validations.js#L40).
This requires some (sometimes tedious) maintenance, but improves the developer experience.

The configuration object is normalized, which includes:

- Assigning
  [default values](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/normalize.js#L22-L32)
- Allowing some properties
  [to be capitalized](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/case.js#L2),
  for backward compatibility
- [Normalizing the Functions-related properties](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/functions_config.js#L11),
  which are rather complex and polymorphic

## Debugging

A `debug` flag can be used
[to log the flags and configuration object](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/log/main.js#L6)
at different stages of the logic. It can enabled with the `NETLIFY_BUILD_DEBUG` environment variable as well.

We make sure that no secret information is ever logged by this debugging feature.
