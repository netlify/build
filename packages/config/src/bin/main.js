#!/usr/bin/env node
const process = require('process')

const { stableStringify } = require('fast-safe-stringify')

require('../utils/polyfills')

const { isUserError } = require('../error')
const resolveConfig = require('../main')

const { parseFlags } = require('./flags')

// CLI entry point
const runCli = async function() {
  try {
    const { stable, ...flags } = parseFlags()
    const result = await resolveConfig(flags)
    handleCliSuccess(result, stable)
  } catch (error) {
    handleCliError(error)
  }
}

// The result is printed as JSON on stdout on success (exit code 0)
const handleCliSuccess = function(result, stable) {
  const resultA = serializeApi(result)
  const stringifyFunc = stable ? stableStringify : JSON.stringify
  const resultJson = stringifyFunc(resultA, null, 2)
  console.log(resultJson)
  process.exitCode = 0
}

// `api` is not JSON-serializable, so we remove it
// We still indicate it as a boolean
const serializeApi = function({ api, ...result }) {
  if (api === undefined) {
    return result
  }

  return { ...result, hasApi: true }
}

const handleCliError = function(error) {
  // Errors caused by users do not show stack traces and have exit code 1
  if (isUserError(error)) {
    console.error(error.message)
    process.exitCode = 1
    return
  }

  // Internal errors / bugs have exit code 2
  console.error(error.stack)
  process.exitCode = 2
}

runCli()
