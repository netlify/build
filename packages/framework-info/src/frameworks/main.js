/* eslint-disable node/global-require,import/order,import/max-dependencies */
// We purposely order the following array to ensure the most relevant framework
// is always first, if several frameworks are detected at once
const FRAMEWORKS = [
  // Static site generators
  require('./docusaurus.json'),
  require('./docusaurus-v2.json'),
  require('./eleventy.json'),
  require('./gatsby.json'),
  require('./gridsome.json'),
  require('./hexo.json'),
  require('./hugo.json'),
  require('./jekyll.json'),
  require('./middleman.json'),
  require('./next.json'),
  require('./nuxt.json'),
  require('./phenomic.json'),
  require('./react-static.json'),
  require('./stencil.json'),
  require('./vuepress.json'),
  require('./assemble.json'),
  require('./docpad.json'),
  require('./harp.json'),
  require('./metalsmith.json'),
  require('./roots.json'),
  require('./wintersmith.json'),

  // Front-end frameworks
  require('./angular.json'),
  require('./create-react-app.json'),
  require('./ember.json'),
  require('./expo.json'),
  require('./quasar.json'),
  require('./quasar-v0.17.json'),
  require('./sapper.json'),
  require('./svelte.json'),
  require('./vue.json'),

  // Build tools
  require('./brunch.json'),
  require('./parcel.json'),
  require('./grunt.json'),
  require('./gulp.json'),
  require('./vite.json'),
]

/* eslint-enable node/global-require,import/order,import/max-dependencies */

module.exports = { FRAMEWORKS }
