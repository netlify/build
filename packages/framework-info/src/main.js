const { getContext } = require('./context')
const { listFrameworks: list, hasFramework: has, getFramework: get } = require('./core.js')

/**
 * @typedef {object} Options
 * @property {string} [projectDir=process.cwd()] - Project's directory
 * @property {string} [nodeVersion=process.version] - Node.js version of the runtime environment. Used to recommend Netlify build plugins
 */

/**
 * @typedef {object} Dev
 * @property {string} commands - Dev command. There might be several alternatives or empty
 * @property {number} port - Server port
 */

/**
 * @typedef {object} Build
 * @property {string} commands - Build command. There might be several alternatives
 * @property {string} directory - Relative path to the directory where files are built
 */

/**
 * @typedef {object} Framework
 * @property {string} id - Id such as `"gatsby"`
 * @property {string} category - Category among `"static_site_generator"`, `"frontend_framework"` and `"build_tool"`
 * @property {Dev} dev - Information about the dev command
 * @property {Build} build - Information about the build command
 * @property {object} env - Environment variables that should be set when calling the dev command
 * @property {string[]} plugins - A list of recommend Netlify build plugins to install for the framework
 */

/**
 * Return all the frameworks used by a project.
 *
 * @param  {Options} options - Options
 *
 * @returns {Framework[]} frameworks - Frameworks used by a project
 */
const listFrameworks = async function (opts) {
  const context = await getContext(opts)
  return await list(context)
}

/**
 * Return whether a project uses a specific framework
 *
 * @param {string} frameworkId - Id such as `"gatsby"`
 * @param  {Options} [options] - Context
 *
 * @returns {boolean} result - Whether the project uses this framework
 */
const hasFramework = async function (frameworkId, options) {
  const context = await getContext(options)
  return await has(frameworkId, context)
}

/**
 * Return some information about a framework used by a project.
 *
 * @param {string} frameworkId - Id such as `"gatsby"`
 * @param  {Context} [context] - Context
 *
 * @returns {Framework} framework - Framework used by a project
 */
const getFramework = async function (frameworkId, options) {
  const context = await getContext(options)
  return await get(frameworkId, context)
}

module.exports = { listFrameworks, hasFramework, getFramework }
