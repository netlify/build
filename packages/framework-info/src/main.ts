import { join } from 'path'
import { cwd } from 'process'

import { findUp } from 'find-up'

import { getContext, getPackageJson } from './context.js'
import { listFrameworks as list, hasFramework as has, getFramework as get } from './core.js'
import type { FrameworkName } from './generated/frameworkNames.js'
import type { Framework } from './types.js'

interface Options {
  projectDir?: string
  nodeVersion?: string
}

/**
 * Gets the version of the framework that is installed in a project.
 *
 * This cannot currently be used in the browser at this time, which is why it's defined
 * here rather than in `core.js` as part of the `getFrameworkInfo` method
 */
const getFrameworkVersion = async (projectDir: string, frameworkInfo: Framework): Promise<Framework> => {
  if (!frameworkInfo.package || !frameworkInfo.package.name) {
    return frameworkInfo
  }

  const npmPackage = frameworkInfo.package.name

  // Get path of package.json for the installed framework. We need to traverse up the directories
  // in the event that the project uses something like npm workspaces, and the installed framework package
  // has been hoisted to the root directory of the project (which differs from the directory of the project/application being built)
  const installedFrameworkPath = await findUp(join('node_modules', npmPackage, 'package.json'), { cwd: projectDir })

  if (!installedFrameworkPath) {
    return frameworkInfo
  }

  const { packageJson } = await getPackageJson(installedFrameworkPath)

  return {
    ...frameworkInfo,
    package: {
      name: npmPackage,
      version: packageJson?.version || 'unknown',
    },
  }
}

/**
 * Return all the frameworks used by a project.
 */
export const listFrameworks = async function (opts: Options = {}): Promise<Framework[]> {
  const context = await getContext(opts.projectDir, opts.nodeVersion)
  const frameworkList = await list(context)

  const projectDir = opts && opts.projectDir ? opts.projectDir : cwd()

  const settledPromises = await Promise.allSettled(
    frameworkList.map((framework) => getFrameworkVersion(projectDir, framework)),
  )
  const updatedList = settledPromises.map((result) => {
    if (result.status === 'fulfilled') {
      return result.value
    }

    throw result.reason
  })

  return updatedList
}

/**
 * Return whether a project uses a specific framework
 */
export const hasFramework = async function (frameworkId: FrameworkName, options: Options = {}): Promise<boolean> {
  const context = await getContext(options.projectDir, options.nodeVersion)
  return has(frameworkId, context)
}

/**
 * Return some information about a framework used by a project.
 */
export const getFramework = async function (frameworkId: FrameworkName, options: Options = {}): Promise<Framework> {
  const context = await getContext(options.projectDir, options.nodeVersion)

  return get(frameworkId, context)
}

export const supportedRuntimes = {
  next: { package: '@netlify/plugin-nextjs', skipFlag: 'NETLIFY_NEXT_PLUGIN_SKIP' },
  gatsby: { package: '@netlify/plugin-gatsby', skipFlag: 'NETLIFY_GATSBY_PLUGIN_SKIP' },
}
