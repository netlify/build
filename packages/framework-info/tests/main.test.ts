import { describe, expect, test } from 'vitest'

import { getFrameworks, getFramework, hasFramework } from './helpers/main.js'

describe('getFrameworks', () => {
  test('Should detect frameworks', async () => {
    const frameworks = await getFrameworks('simple')
    expect(frameworks).toMatchSnapshot()
  })

  test('Should return an empty array when no framework is detected', async () => {
    const frameworks = await getFrameworks('empty')
    expect(frameworks).toHaveLength(0)
  })

  test('Should return several items when multiple frameworks are detected', async () => {
    const frameworks = await getFrameworks('multiple')
    expect(frameworks).toHaveLength(2)
  })

  test('Should return the version of each framework when multiple are detected', async () => {
    const frameworks = await getFrameworks('multiple')
    expect(frameworks).toMatchSnapshot()
  })

  test('Should return the version of a framework that is not detected by npm package', async () => {
    const frameworks = await getFrameworks('no_package')
    expect(frameworks).toMatchSnapshot()
  })

  test('Should return the version of the framework when the installed package is hoisted to the root project directory', async () => {
    const frameworks = await getFrameworks('monorepos/app1')
    expect(frameworks).toMatchSnapshot()
  })

  test('Should work if version cannot be detected', async () => {
    const frameworks = await getFrameworks('no-version')
    expect(frameworks).toMatchSnapshot()
  })
})

describe('getFramework', () => {
  test('Should allow getting a specific framework', async () => {
    const framework = await getFramework('simple', 'sapper')
    expect(framework).toMatchSnapshot()
  })

  test('Should throw when passing an invalid framework', async () => {
    await expect(getFramework('simple', 'doesNotExist')).rejects.toThrowError()
  })
})

describe('hasFramework', () => {
  test('Should allow testing a specific framework', async () => {
    expect(await hasFramework('simple', 'sapper')).toBe(true)
    expect(await hasFramework('simple', 'nuxt')).toBe(false)
  })

  test('Should throw when testing an invalid framework', async () => {
    await expect(hasFramework('simple', 'doesNotExist')).rejects.toThrowError()
  })

  test('Should sort framework ids in invalid framework error message', async () => {
    try {
      await hasFramework('simple', 'doesNotExist')
      expect.fail('should throw')
    } catch (error) {
      const [, frameworksFromMessage] = error.message.match(/It should be one of: (.+)/)
      const frameworksArray = frameworksFromMessage.split(', ')

      expect(frameworksArray).toEqual([...frameworksArray].sort())
    }
  })
})
