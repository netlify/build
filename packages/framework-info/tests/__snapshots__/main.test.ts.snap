// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getFramework > Should allow getting a specific framework 1`] = `
{
  "build": {
    "commands": [
      "sapper export",
    ],
    "directory": "__sapper__/export",
  },
  "category": "frontend_framework",
  "dev": {
    "commands": [
      "sapper dev",
    ],
    "pollingStrategies": [
      {
        "name": "TCP",
      },
      {
        "name": "HTTP",
      },
    ],
    "port": 3000,
  },
  "env": {},
  "id": "sapper",
  "logo": {
    "dark": "https://framework-info.netlify.app/logos/sapper/default.svg",
    "default": "https://framework-info.netlify.app/logos/sapper/default.svg",
    "light": "https://framework-info.netlify.app/logos/sapper/default.svg",
  },
  "name": "Sapper",
  "package": {
    "name": "sapper",
    "version": "unknown",
  },
  "plugins": [],
  "staticAssetsDirectory": "static",
}
`;

exports[`getFrameworks > Should detect frameworks 1`] = `
[
  {
    "build": {
      "commands": [
        "sapper export",
      ],
      "directory": "__sapper__/export",
    },
    "category": "frontend_framework",
    "dev": {
      "commands": [
        "sapper dev",
      ],
      "pollingStrategies": [
        {
          "name": "TCP",
        },
        {
          "name": "HTTP",
        },
      ],
      "port": 3000,
    },
    "env": {},
    "id": "sapper",
    "logo": {
      "dark": "https://framework-info.netlify.app/logos/sapper/default.svg",
      "default": "https://framework-info.netlify.app/logos/sapper/default.svg",
      "light": "https://framework-info.netlify.app/logos/sapper/default.svg",
    },
    "name": "Sapper",
    "package": {
      "name": "sapper",
      "version": "3.4.3",
    },
    "plugins": [],
    "staticAssetsDirectory": "static",
  },
]
`;

exports[`getFrameworks > Should return the version of a framework that is not detected by npm package 1`] = `
[
  {
    "build": {
      "commands": [
        "bundle exec middleman build",
      ],
      "directory": "build",
    },
    "category": "static_site_generator",
    "dev": {
      "commands": [
        "bundle exec middleman server",
      ],
      "pollingStrategies": [
        {
          "name": "TCP",
        },
        {
          "name": "HTTP",
        },
      ],
      "port": 4567,
    },
    "env": {},
    "id": "middleman",
    "logo": {
      "dark": "https://framework-info.netlify.app/logos/middleman/default.svg",
      "default": "https://framework-info.netlify.app/logos/middleman/default.svg",
      "light": "https://framework-info.netlify.app/logos/middleman/default.svg",
    },
    "name": "Middleman",
    "package": {
      "name": undefined,
      "version": "unknown",
    },
    "plugins": [],
    "staticAssetsDirectory": undefined,
  },
]
`;

exports[`getFrameworks > Should return the version of each framework when multiple are detected 1`] = `
[
  {
    "build": {
      "commands": [
        "vuepress build",
      ],
      "directory": ".vuepress/dist",
    },
    "category": "static_site_generator",
    "dev": {
      "commands": [
        "vuepress dev",
      ],
      "pollingStrategies": [
        {
          "name": "TCP",
        },
        {
          "name": "HTTP",
        },
      ],
      "port": 8080,
    },
    "env": {},
    "id": "vuepress",
    "logo": {
      "dark": "https://framework-info.netlify.app/logos/vuepress/default.svg",
      "default": "https://framework-info.netlify.app/logos/vuepress/default.svg",
      "light": "https://framework-info.netlify.app/logos/vuepress/default.svg",
    },
    "name": "VuePress",
    "package": {
      "name": "vuepress",
      "version": "4.5.6",
    },
    "plugins": [],
    "staticAssetsDirectory": undefined,
  },
  {
    "build": {
      "commands": [
        "vue-cli-service build",
      ],
      "directory": "dist",
    },
    "category": "frontend_framework",
    "dev": {
      "commands": [
        "vue-cli-service serve",
      ],
      "pollingStrategies": [
        {
          "name": "TCP",
        },
        {
          "name": "HTTP",
        },
      ],
      "port": 8080,
    },
    "env": {},
    "id": "vue",
    "logo": {
      "dark": "https://framework-info.netlify.app/logos/vue/default.svg",
      "default": "https://framework-info.netlify.app/logos/vue/default.svg",
      "light": "https://framework-info.netlify.app/logos/vue/default.svg",
    },
    "name": "Vue.js",
    "package": {
      "name": "@vue/cli-service",
      "version": "1.2.3",
    },
    "plugins": [],
    "staticAssetsDirectory": undefined,
  },
]
`;

exports[`getFrameworks > Should return the version of the framework when the installed package is hoisted to the root project directory 1`] = `
[
  {
    "build": {
      "commands": [
        "next build",
      ],
      "directory": ".next",
    },
    "category": "static_site_generator",
    "dev": {
      "commands": [
        "next",
      ],
      "pollingStrategies": [
        {
          "name": "TCP",
        },
      ],
      "port": 3000,
    },
    "env": {},
    "id": "next",
    "logo": {
      "dark": "https://framework-info.netlify.app/logos/nextjs/dark.svg",
      "default": "https://framework-info.netlify.app/logos/nextjs/light.svg",
      "light": "https://framework-info.netlify.app/logos/nextjs/light.svg",
    },
    "name": "Next.js",
    "package": {
      "name": "next",
      "version": "3.2.1",
    },
    "plugins": [
      "@netlify/plugin-nextjs",
    ],
    "staticAssetsDirectory": undefined,
  },
]
`;

exports[`getFrameworks > Should work if version cannot be detected 1`] = `
[
  {
    "build": {
      "commands": [
        "docpad generate",
      ],
      "directory": "out",
    },
    "category": "static_site_generator",
    "dev": {
      "commands": [
        "docpad run",
      ],
      "pollingStrategies": [
        {
          "name": "TCP",
        },
        {
          "name": "HTTP",
        },
      ],
      "port": 9778,
    },
    "env": {},
    "id": "docpad",
    "logo": undefined,
    "name": "DocPad",
    "package": {
      "name": "docpad",
      "version": "unknown",
    },
    "plugins": [],
    "staticAssetsDirectory": undefined,
  },
]
`;
